zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -306,6 +669,431 @@ impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+                    (
+                        q_sel.expr()
+                            * meta.query_fixed(block_table_index_blockhash, Rotation::cur()),
+                        meta.query_advice(block_table_blockhash.index, Rotation::cur()),
+                    ),
+                    (
+                        q_sel.expr()
+                            * meta.query_advice(blk_hdr_reconstruct_value, Rotation::cur()),
+                        meta.query_advice(block_table_blockhash.value, Rotation::cur()),
+                    ),
+                ]
+            },
+        );
+
+        // 3. Check block header hash
+        meta.lookup_any("blockhash lookup keccak", |meta| {
+            let q_blk_hdr_rlp_end = meta.query_selector(q_blk_hdr_rlp_end);
+            let blk_hdr_rlc = meta.query_advice(blk_hdr_rlc_acc, Rotation::cur());
+            // The total RLP length is the RLP list length (0x200 + blk_hdr_rlp[2])
+            //  + 3 bytes for the RLP list header
+            let blk_hdr_rlp_num_bytes = 0x200.expr()
+                + meta.query_advice(
+                    blk_hdr_rlp,
+                    Rotation(-(BLOCKHASH_TOTAL_ROWS as i32) + 1 + 2),
+                )
+                + 3.expr();
+            let blk_hdr_hash_hi = meta.query_advice(blk_hdr_blockhash, Rotation::cur());
+            let blk_hdr_hash_lo = meta.query_advice(blk_hdr_blockhash, Rotation::prev());
+            vec![
+                (
+                    q_blk_hdr_rlp_end.expr(),
+                    meta.query_advice(keccak_table2.is_enabled, Rotation::cur()),
+                ),
+                (
+                    q_blk_hdr_rlp_end.expr() * blk_hdr_rlc,
+                    meta.query_advice(keccak_table2.input_rlc, Rotation::cur()),
+                ),
+                (
+                    q_blk_hdr_rlp_end.expr() * blk_hdr_rlp_num_bytes,
+                    meta.query_advice(keccak_table2.input_len, Rotation::cur()),
+                ),
+                (
+                    q_blk_hdr_rlp_end.expr() * blk_hdr_hash_hi,
+                    meta.query_advice(keccak_table2.output_hi, Rotation::cur()),
+                ),
+                (
+                    q_blk_hdr_rlp_end * blk_hdr_hash_lo,
+                    meta.query_advice(keccak_table2.output_lo, Rotation::cur()),
+                ),
+            ]
+        });
+
+
+        for rotation in [0,-1] { // hi blockhash is in current offset, lo blockhash in in previous
+            meta.lookup_any(
+                "Block header: Check hi/lo parts of block hashes against previous hashes",
+                |meta| {
+                    let q_blk_hdr_rlp_end = meta.query_selector(q_blk_hdr_rlp_end);
+                    let blk_hdr_hash_hi = meta.query_advice(blk_hdr_blockhash, Rotation(rotation));
+                    let q_lookup_blockhash = meta.query_selector(q_lookup_blockhash);
+                    let tag = meta.query_fixed(block_table_tag_blockhash, Rotation(rotation-1));
```

Comment:
Does tag rotation need to be -1 & -2?? looks a little bit strange.

Reply:
They cannot be in the same row because this is where `withdrawals_root` tag also needs to be. I could put it in a different column but I didn't want to add a col for only just this case.